### ISSUE TITLE
serving non-existent favicon.ico with staticfile throws error on refresh

### ISSUE THREAD

**Issue by ghost** [cherrypy/cherrypy#983]
Originally reported by: **@wyuenho**

---

So apparently Tree.mount() attaches a static handler for favicon.ico if the root doesn't already have a handler for it without even looking at the config. If the config has a section for /favicon.ico, but the file doesn't exist, _refreshing_ the browser will throw and error in the errorlog with something like this:

```
[17/Jan/2010:01:30:00] HTTP Traceback (most recent call last):
  File "/Users/wyuenho/Documents/spiralup/spiralup/lib/python2.5/site-packages/CherryPy-3.1.2-py2.5.egg/cherrypy/_cprequest.py", line 606, in respond
    cherrypy.response.body = self.handler()
  File "/Users/wyuenho/Documents/spiralup/spiralup/lib/python2.5/site-packages/CherryPy-3.1.2-py2.5.egg/cherrypy/_cpdispatch.py", line 25, in __call__
    return self.callable(*self.args, **self.kwargs)
  File "/Users/wyuenho/Documents/spiralup/spiralup/lib/python2.5/site-packages/CherryPy-3.1.2-py2.5.egg/cherrypy/_cptools.py", line 150, in handle_func
    handled = self.callable(*args, **self._merged_args(kwargs))
TypeError: staticfile() got multiple values for keyword argument 'filename'
```

This obviously doesn't even tell me what values it got, so the error message doesn't even help. I'm not sure how to patch this because you need to obviously look at all the different kinds of dispatchers for the "/favicon.ico" key besides the default tree dispatcher, and I'm not familiar with all of them.

Reported by wyuenho@gmail.com

---
- Bitbucket: https://bitbucket.org/cherrypy/cherrypy/issue/983

**Comment by Safihre** [comment]
Error is also occurring for some of our users.
Is there a way to avoid this? We specify a favicon in the config for the root `'/favicon.ico'`.
From the logs it happens specifically when they get port-scanned by some tools.. But can't seem to get from the log what URL it is requesting.

**Comment by Safihre** [comment]
Error-log:

```
2016-10-05 10:28:17,203::INFO::[_cplogging:217] [05/Oct/2016:10:28:17] HTTP 
Request Headers:
  Content-Length: 0
  PRAGMA: no-cache
  HOST: t19.proxy-checks.com
  Remote-Addr: 95.25.212.145
  ACCEPT: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8
  USER-AGENT: Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 6.0; SLCC1; .NET CLR 2.0.50727; Media Center PC 5.0; .NET CLR 3.5.21022; .NET CLR 3.5.30729; .NET CLR 3.0.30618)
  PROXY-CONNECTION: Keep-Alive
  ACCEPT-LANGUAGE: en-US;q=0.6,en;q=0.4
2016-10-05 10:28:17,203::ERROR::[_cplogging:217] [05/Oct/2016:10:28:17] HTTP 
Traceback (most recent call last):
  File "cherrypy\_cprequest.pyo", line 670, in respond
  File "cherrypy\lib\encoding.pyo", line 220, in __call__
  File "cherrypy\_cpdispatch.pyo", line 60, in __call__
  File "cherrypy\_cptools.pyo", line 177, in handle_func
TypeError: staticfile() got multiple values for keyword argument 'filename'
```

**Comment by webknjaz** [comment]
@wyuenho

**Comment by webknjaz** [comment]
@Safihre  Is this reproducible for currently supported version?

**Comment by Safihre** [comment]
This was on 6.0.2 (older SABnzbd release).
But not sure how to reproduce it.. Trying.

**Comment by Safihre** [comment]
@webknjaz Can reproduce it now:
It happens when the favicon is part of a `tools.staticdir`.
Example:
```python
{
 '/staticcfg' : {'tools.staticdir.on': True, 'tools.staticdir.dir': '/bla/staticcfg/' }
 '/favicon.ico': {'tools.staticfile.on': True, 'tools.staticfile.filename': '/bla/staticcfg/favicon.ico'}
}
```

**Comment by Safihre** [comment]
Actually I was wrong. I was linking to a non-existing file. Will try to reproduce again.

**Comment by blshkv** [comment]
I got the same message after path mapping using ```staticfile.filename`` to an invalid location. The POC code is below:
```
    cherrypy.tree.mount(Root(), '/', {                                                                               
        '/favicon.ico' : {                                                                                           
            'tools.staticfile.on' : True,                                                                            
            'tools.staticfile.filename' : ROOT + '/imgs/favicon.ico'
```
The file ```/imgs/favicon.ico``` was missing

Perhaps, you can make the error clearer?

**Comment by webknjaz** [comment]
So this is what's happening:
```python
> ~/src/github/cherrypy/cherrypy/cherrypy/_cptools.py(183)handle_func()
    182         def handle_func(*a, **kw):
--> 183             handled = self.callable(*args, **self._merged_args(kwargs))
    184             if not handled:

ipdb> p args
('~/src/github/cherrypy/cherrypy/cherrypy/favicon.ico',)
ipdb> self._merged_args(kwargs)
{'filename': '/dev/null/x'}
```

Still, we need to find the root of this issue.
Call stack looks as follows:
```python-traceback
ipdb> import traceback                                                                                                                                         [33/330]
ipdb> traceback.print_stack()
  File "~/.pyenv/versions/3.7.0/lib/python3.7/threading.py", line 885, in _bootstrap
    self._bootstrap_inner()
  File "~/.pyenv/versions/3.7.0/lib/python3.7/threading.py", line 917, in _bootstrap_inner
    self.run()
  File "~/src/github/cherrypy/cheroot/cheroot/workers/threadpool.py", line 112, in run
    conn.communicate()
  File "~/src/github/cherrypy/cheroot/cheroot/server.py", line 1214, in communicate
    req.respond()
  File "~/src/github/cherrypy/cheroot/cheroot/server.py", line 1018, in respond
    self.server.gateway(self).respond()
  File "~/src/github/cherrypy/cheroot/cheroot/wsgi.py", line 142, in respond
    response = self.req.server.wsgi_app(self.env, self.start_response)
  File "~/src/github/cherrypy/cherrypy/cherrypy/_cptree.py", line 313, in __call__
    return app(environ, start_response)
  File "~/src/github/cherrypy/cherrypy/cherrypy/_cptree.py", line 171, in __call__
    return self.wsgiapp(environ, start_response)
  File "~/src/github/cherrypy/cherrypy/cherrypy/_cpwsgi.py", line 451, in __call__
    return head(environ, start_response)
  File "~/src/github/cherrypy/cherrypy/cherrypy/_cpwsgi.py", line 154, in __call__
    self.throws
  File "~/src/github/cherrypy/cherrypy/cherrypy/_cpwsgi.py", line 169, in __init__
    self.nextapp, self.environ, self.start_response,
  File "~/src/github/cherrypy/cherrypy/cherrypy/_cpwsgi.py", line 190, in trap
    return func(*args, **kwargs)
  File "~/src/github/cherrypy/cherrypy/cherrypy/_cpwsgi.py", line 106, in __call__
    return self.nextapp(environ, start_response)
  File "~/src/github/cherrypy/cherrypy/cherrypy/_cpwsgi.py", line 439, in tail
    return self.response_class(environ, start_response, self.cpapp)
  File "~/src/github/cherrypy/cherrypy/cherrypy/_cpwsgi.py", line 245, in __init__
    self.run()
  File "~/src/github/cherrypy/cherrypy/cherrypy/_cpwsgi.py", line 348, in run
    request.run(meth, path, qs, rproto, headers, rfile)
  File "~/src/github/cherrypy/cherrypy/cherrypy/_cprequest.py", line 593, in run
    self.respond(pi)
  File "~/src/github/cherrypy/cherrypy/cherrypy/_cprequest.py", line 627, in respond
    self._do_respond(path_info)
  File "~/src/github/cherrypy/cherrypy/cherrypy/_cprequest.py", line 686, in _do_respond
    response.body = self.handler()
  File "~/src/github/cherrypy/cherrypy/cherrypy/lib/encoding.py", line 221, in __call__
    self.body = self.oldhandler(*args, **kwargs)
  File "~/src/github/cherrypy/cherrypy/cherrypy/_cpdispatch.py", line 54, in __call__
    return self.callable(*self.args, **self.kwargs)
  File "~/src/github/cherrypy/cherrypy/cherrypy/_cptools.py", line 183, in handle_func
    handled = self.callable(*args, **self._merged_args(kwargs))
  File "~/.pyenv/versions/3.7.0/lib/python3.7/bdb.py", line 88, in trace_dispatch
    return self.dispatch_line(frame)
  File "~/.pyenv/versions/3.7.0/lib/python3.7/bdb.py", line 112, in dispatch_line
    self.user_line(frame)
  File "~/.pyenv/versions/3.7.0/lib/python3.7/pdb.py", line 261, in user_line
    self.interaction(frame, None)
  File "~/.pyenv/versions/3.7.0/lib/python3.7/site-packages/IPython/core/debugger.py", line 297, in interaction
    OldPdb.interaction(self, frame, traceback)
  File "~/.pyenv/versions/3.7.0/lib/python3.7/pdb.py", line 352, in interaction
    self._cmdloop()
  File "~/.pyenv/versions/3.7.0/lib/python3.7/pdb.py", line 321, in _cmdloop
    self.cmdloop()
  File "~/.pyenv/versions/3.7.0/lib/python3.7/site-packages/IPython/terminal/debugger.py", line 99, in cmdloop
    stop = self.onecmd(line)
  File "~/.pyenv/versions/3.7.0/lib/python3.7/pdb.py", line 418, in onecmd
    return cmd.Cmd.onecmd(self, line)
  File "~/.pyenv/versions/3.7.0/lib/python3.7/cmd.py", line 216, in onecmd
    return self.default(line)
  File "~/.pyenv/versions/3.7.0/lib/python3.7/pdb.py", line 376, in default
    exec(code, globals, locals)
  File "<stdin>", line 1, in <module>
```

**Comment by webknjaz** [comment]
It's probably related to https://github.com/cherrypy/cherrypy/blob/03cf03b/cherrypy/_cptree.py#L236-L248:
```python
            # If mounted at "", add favicon.ico
            needs_favicon = (
                script_name == ''
                and root is not None
                and not hasattr(root, 'favicon_ico')
            )
            if needs_favicon:
                favicon = os.path.join(
                    os.getcwd(),
                    os.path.dirname(__file__),
                    'favicon.ico',
                )
                root.favicon_ico = tools.staticfile.handler(favicon)
```

**Comment by webknjaz** [comment]
Reproducer:
```python
#! /usr/bin/env python
import cherrypy

conf = {
    '/favicon.ico': {
        'tools.staticfile.on': True,
        'tools.staticfile.filename': '/dev/null/x',
    }
}


class HelloWorld:
    @cherrypy.expose
    def index(self, **params):
        return "Hello world!"


if __name__ == '__main__':
    cherrypy.quickstart(HelloWorld(), '/', conf)
```

**Comment by sanderjo** [comment]
That indeed reproduces it

```
  File "/usr/local/lib/python2.7/dist-packages/cherrypy/_cptools.py", line 182, in handle_func
    handled = self.callable(*args, **self._merged_args(kwargs))
TypeError: staticfile() got multiple values for keyword argument 'filename'
```
With the monkey-patch from https://github.com/cherrypy/cherrypy/issues/1623#issue-250899142 (which I don't understand at all) , the traceback is gone.

```
sander@Stream-13:~$ python cherrypy-reproducer.py 
[21/Jul/2018:17:31:39] ENGINE Listening for SIGHUP.
[21/Jul/2018:17:31:39] ENGINE Listening for SIGTERM.
[21/Jul/2018:17:31:39] ENGINE Listening for SIGUSR1.
[21/Jul/2018:17:31:39] ENGINE Bus STARTING
[21/Jul/2018:17:31:39] ENGINE Started monitor thread 'Autoreloader'.
[21/Jul/2018:17:31:39] ENGINE Serving on http://127.0.0.1:8080
[21/Jul/2018:17:31:39] ENGINE Bus STARTED
127.0.0.1 - - [21/Jul/2018:17:31:43] "GET / HTTP/1.1" 200 12 "" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.59 Safari/537.36"
127.0.0.1 - - [21/Jul/2018:17:31:43] "GET /favicon.ico HTTP/1.1" 200 1406 "http://127.0.0.1:8080/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.59 Safari/537.36"
```

Monkey-patch:
```
root@Stream-13:~# diff /usr/local/lib/python2.7/dist-packages/cherrypy/_cptools.py.ORG /usr/local/lib/python2.7/dist-packages/cherrypy/_cptools.py
182c182,189
<             handled = self.callable(*args, **self._merged_args(kwargs))
---
>             # handled = self.callable(*args, **self._merged_args(kwargs))
> 
>             parameters = ('filename', 'root', 'match', 'content_types', 'debug')
>             kws = self._merged_args(kwargs)
>             for i in range(len(args)):
>                 kws.pop(parameters[i], None) 
>             handled = self.callable(*args, **kws)
> 
```




Full traceback **without** the monkey-patch:

```
sander@Stream-13:~$ python cherrypy-reproducer.py 
[21/Jul/2018:17:25:39] ENGINE Listening for SIGHUP.
[21/Jul/2018:17:25:39] ENGINE Listening for SIGTERM.
[21/Jul/2018:17:25:39] ENGINE Listening for SIGUSR1.
[21/Jul/2018:17:25:39] ENGINE Bus STARTING
[21/Jul/2018:17:25:39] ENGINE Started monitor thread 'Autoreloader'.
[21/Jul/2018:17:25:39] ENGINE Serving on http://127.0.0.1:8080
[21/Jul/2018:17:25:39] ENGINE Bus STARTED
127.0.0.1 - - [21/Jul/2018:17:25:50] "GET / HTTP/1.1" 200 12 "" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.59 Safari/537.36"
[21/Jul/2018:17:25:52] HTTP 
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/cherrypy/_cprequest.py", line 627, in respond
    self._do_respond(path_info)
  File "/usr/local/lib/python2.7/dist-packages/cherrypy/_cprequest.py", line 686, in _do_respond
    response.body = self.handler()
  File "/usr/local/lib/python2.7/dist-packages/cherrypy/lib/encoding.py", line 221, in __call__
    self.body = self.oldhandler(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/cherrypy/_cpdispatch.py", line 54, in __call__
    return self.callable(*self.args, **self.kwargs)
  File "/usr/local/lib/python2.7/dist-packages/cherrypy/_cptools.py", line 182, in handle_func
    handled = self.callable(*args, **self._merged_args(kwargs))
TypeError: staticfile() got multiple values for keyword argument 'filename'
[21/Jul/2018:17:25:52] HTTP 
Request Headers:
  REFERER: http://127.0.0.1:8080/
  HOST: 127.0.0.1:8080
  CONNECTION: keep-alive
  Remote-Addr: 127.0.0.1
  ACCEPT: image/webp,image/apng,image/*,*/*;q=0.8
  USER-AGENT: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.59 Safari/537.36
  ACCEPT-LANGUAGE: en-US,en;q=0.9,nl;q=0.8,de;q=0.7,fr;q=0.6
  ACCEPT-ENCODING: gzip, deflate, br
127.0.0.1 - - [21/Jul/2018:17:25:52] "GET /favicon.ico HTTP/1.1" 500 1642 "http://127.0.0.1:8080/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.59 Safari/537.36"
```

**Comment by webknjaz** [comment]
> With the monkey-patch from #1623 (comment) (which I don't understand at all) , the traceback is gone.

That patch removes the duplicate argument. But it "cures" the symptom, not an underlying issue. It looks like the fix should be done on a different level, being more fundamental. The patch might hide other problems if they happen, which would then be impossible to debug.

To truly fix the issue, we need to understand the flow across different components of the framework and why this is happening. Otherwise patching random places would be harmful.

Basically, this is why I postponed the fix, it needs mode dedicated investigation first.

**Comment by webknjaz** [comment]
In #1623 @sanjaySP suggested to register a custom path for favicon and explicitly specify that in your HTML pages *as a workaround*.
